### ISSUE TITLE
Add CMake config option to disable installing

### ISSUE THREAD

**Issue by VelocityRa** [jbeder/yaml-cpp#628]
Originally suggested in #624, made into a separate issue to prevent confusion.

**Comment by VelocityRa** [comment]
To be honest, I'm not even sure what "installing" means in this context. Yaml-cpp is statically linked with my application and the only `main` I can find in this project, is for tests.
Is it "installing" like how you'd install a library from a package manager on Linux? Like, compiled libs and headers?

**Comment by stotko** [comment]
Yes, it means that you install the compiled library, headers, and additional CMake files at the location defined by `CMAKE_INSTALL_PREFIX`. If you deploy your project, i.e. installing it or packing it together for later installation at some client, then I would also deploy the subprojects since the main project directly depends on that particular version.

The only use case I can think of, where installing the subproject is not necessary or needed, is when you can ensure that the client has at least some ABI compatible version installed. But this seems extremely error-prone and complicated compared to the default strategy. One exception is of course in case unit tests.

**Comment by kukkamario** [comment]
@stotko Installing yaml-cpp with a project (lets call is 'A') makes sense when yaml-cpp is "public dependency" meaning that headers of yaml-cpp are included from A's the public headers. This causes everyone that uses A's headers to also need yaml-cpp in the include path and link to yaml-cpp libraries. But if yaml-cpp is linked statically and only used inside A's source files and private headers or the project A is an executable project instead of a library, then anyone using A, doesn't need yaml-cpp. yaml-cpp is then "private dependency" to the project A.

In our use case, the project A is an executable. User of the A doesn't need headers or libraries of yaml-cpp because they only want to use the project A and A doesn't require yaml-cpp after the build process (static linking). Usually no one wants to install anything unnecessary and it is a bit unexpected if ´make install´ installs some internal dependency.

**Comment by VelocityRa** [comment]
Thanks, that cleared it up.

So it sounds like the default should be to _not_ install. Probably still have it as an option, though?
I'm not sure if yaml-cpp should be able to be packaged as @stotko mentioned, I was too under the impression that it can only ever a build-time dependency (static library) as @kukkamario said.

**Comment by stotko** [comment]
I agree that if used as a private dependency and built as a static libary, then allowing to disable installation will make the deployment process a bit cleaner. In all other cases, i.e. either if yaml-cpp is a public dependency or build as a shared library (there is already an option for this), one should install it as it's strictly required.

So I'm fine with the PR as-is where the default value for installing is `ON`. Disabling installation per default, i.e. set the option to `OFF`, will very likely break existing projects relying on the current behavior, so the default should definitely be `ON`.

**Comment by kukkamario** [comment]
I also think that installation should be enabled by default. Usually libraries are anyway installed as a separate projects where installation target generation by default is always what is wanted. CMake subdirectory is a special case.

**Comment by bjfar** [comment]
I am also encountering this behaviour and having it mess up my project. My project first checks if there is a system yaml-cpp installation. If it is available, great, it uses that. But if not, it then builds yaml-cpp itself. But I really don't want this "internal" yaml-cpp to get installed onto the users system. This would just be very unexpected for them. It also screws up my particular build scenario because I have to re-run CMake a couple of times for other reasons, and having yaml-cpp install itself means that it gets picked up as "the system" yaml-cpp installation on the second CMake run, which is causing some odd behaviour because it gets seen different on the two CMake runs. Maybe there is something sub-optimal in my built setup which could be fixed to avoid this, but I also find it very odd for any project to install itself without me explicitly asking it to do so.

**Comment by SGSSGene** [comment]
With PR #625 merged I consider this issue closed.
Feel free to open this ticket again or open a new one if I have missed something.
